src/solver/conv_winoRxS.cpp

```diff
@@ -410,118 +410,205 @@ ConvBinWinoRxS<Winodata, Winofilter>::Search(const ConvolutionContext& ctx,
 }
 
-template <int Winodata, int Winofilter>
 class ShaderModel : public UnifiedDescriptionConv2d
 {
-    const size_t DATATYPE_BITS;    // S
-    const size_t n_groups;         // BQ ~compute units
-    const bool out_of_model_scope; // Shader model produces unreliable results.
+    static constexpr size_t NHW_tiles_factor = 32;
+
+    size_t n_CU;
+    size_t n_groups;
+
+    double direct_convolution_macs;
+
+    size_t R_loops;
+    size_t S_loops;
+    size_t C_loops;
+
+    size_t n_works_per_CU;
+
+    bool dstride2;
+    bool ostride2;
+
+    bool single_mode;
+
+    bool is_fp16;
+    bool is_2x3;
+
+    bool out_of_model_scope; // Shader model produces unreliable results.
 
 public:
-    ShaderModel(const ConvolutionContext& ctx, const ProblemDescription& problem)
+    ShaderModel(const ConvolutionContext& ctx,
+                const ProblemDescription& problem,
+                size_t Winodata,
+                size_t Winofilter)
         : UnifiedDescriptionConv2d(problem),
-          DATATYPE_BITS(problem.IsFp16() ? 16 : 32),
-          n_groups(ctx.GetStream()
-                       .GetMaxHardwareComputeUnits()), /// \todo Take n_groups from PerfConfig.
-          out_of_model_scope(!(problem.GetGroupCount() == 1) || //
-                             !(U == 1) ||                       //
-                             !(V == 1) ||                       //
-                             !(input_stride_h == 1) ||          //
-                             !(input_stride_w == 1) ||          //
-                             !(filter_stride_h == 1) ||         //
-                             !(filter_stride_w == 1) ||         //
+          n_CU{ctx.GetStream().GetMaxHardwareComputeUnits()}, /// \todo Take n_CU from PerfConfig.
+          direct_convolution_macs{static_cast<double>(C * N * K) / 1e+6 *
+                                  static_cast<double>(Ceil(S * out_w, input_stride_w) *
+                                                      Ceil(R * out_h, input_stride_h))},
+          is_fp16{problem.IsFp16()},
+          is_2x3{IS2X3},
+          out_of_model_scope
+    {
+        !(problem.GetGroupCount() == 1) || //
+            !(U == 1) ||                   //
+            !(V == 1) ||                   //
+            !(input_stride_h == 1) ||      //
+            !(input_stride_w == 1) ||      //
+            !(filter_stride_h == 1) ||     //
+            !(filter_stride_w == 1) ||     //
 #if !WTI_MODEL_ALLOW_ANY_RS
-                             !(R <= 5) || //
-                             !(S <= 5) || //
+            !(R <= 5) || //
+            !(S <= 5) || //
 #endif
-                             !(C >= 16) || //
-                             !(K >= 16))
+            !(C >= 16) || //
+            !(K >= 16)
+    }
     {
         // Computations do not support negative padding.
         // Negative padding is not applicable, so let use simple assert here.
         assert(pad_h >= 0 && pad_w >= 0);
```

Comment:
I believe this assert is not needed. Only the padding parity (even or odd) affects granularity.

Reply:
We could remove it